### PR TITLE
Adding initial support for GCP

### PR DESCRIPTION
**Pull request by beraldoleal**
This is basically Cfir' work with me changing paths and small things, to accommodate repository layout changes a few months ago.

We had this on hold because I was working on GKE support as well. However, GKE right now depends on https://github.com/confidential-containers/cloud-api-adaptor/pull/1920. So my proposal is to merge this, since it will not affect other providers, and when #1920 is merged I will send another PR with the GKE support.

### CODE REVIEW

**Comment by wainersm** [comment]
Hi @beraldoleal !

I've already started my review but I'd like to send a few words before making comments in code:
 * I'm not an expert on GCP, actually I barely know it. So always be suspicious about my comments .
 * I'm largely comparing this implementation with Azure and AWS providers.
 * I assume this implementation is a MVP, so some features (e.g. confidential VMs) are missing on purpose. Even though, I may make some comments so that we start building a gap list (if you haven't such as list already). These comments shouldn't block the merge of this work.
 * I may have some difficult to test it entirely as I don´t  have a GCP account (still looking for one). I´ll try my best.

File: src/cloud-providers/gcp/provider.go
Comment (wainersm) [inline src/cloud-providers/gcp/provider.go:126]: IIUC this will create a VM with public IP address. The AWS implementation creates a VM with private IP by default but it allow users to choose to use public IP (see commit 39d69d494b982f008e0b1767469cf914b1fa835f). Other providers I don't know how it was implemented.

File: src/cloud-providers/gcp/provider.go
Comment (wainersm) [inline src/cloud-providers/gcp/provider.go:138]: In case `Insert()` fails, will it automatically delete any resources (e.g. disks and networks) that were created? Asking because Azure does need to delete those resources explicitly, whereas other providers don't.

File: src/cloud-providers/gcp/provider.go
Comment (wainersm) [inline src/cloud-providers/gcp/provider.go:72]: Suggestion: a meaningful log message, e.g., "Found pod node IP: ...."

File: src/cloud-providers/gcp/provider.go
Comment (wainersm) [inline src/cloud-providers/gcp/provider.go:127]: When I read "premium" my brain automatically associates with "expensive". So it would be nice to have a option for users to tune this.

File: src/cloud-providers/gcp/provider.go
Comment (wainersm) [inline src/cloud-providers/gcp/provider.go:92]: You will need to improve the InstanceRequest to allow users to customize the VM. One concept that almost all providers implemented is of "flexible instance sizes" (please see https://github.com/confidential-containers/cloud-api-adaptor/issues/321).

Enable/disable confidential VM is also something desirable that I mentioned elsewhere.

File: src/cloud-providers/gcp/provider.go
Comment (wainersm) [inline src/cloud-providers/gcp/provider.go:189]: Don't know what's the use of `Teardown()`, do you? Only Vsphere provider implements this...

File: src/peerpod-ctrl/Makefile
Comment (wainersm) [inline src/peerpod-ctrl/Makefile:66]: Bonus if you re-order alphabetically :)

File: src/cloud-providers/gcp/provider.go
Comment (wainersm) [inline src/cloud-providers/gcp/provider.go:95]: Another missing feature is the ability for users to specify custom tags (see https://github.com/confidential-containers/cloud-api-adaptor/issues/1280)

File: src/cloud-api-adaptor/entrypoint.sh
Comment (wainersm) [inline src/cloud-api-adaptor/entrypoint.sh:93]: Does GCP has an equivalent to AWS's IMDS?

AWS implements https://github.com/confidential-containers/cloud-api-adaptor/issues/315 which will get information like the subnetId from the IMDS. This means that the subnetId parameter is optional.

Is GCP_PROJECT_ID and GCP_ZONE automatically filled if the user doesn't set? If not, then these variables should be listed in `test_vars` because they become mandatory.

`PODVM_IMAGE_NAME` should be mandatory, thus in `test_vars`.

File: src/cloud-api-adaptor/entrypoint.sh
Comment (wainersm) [inline src/cloud-api-adaptor/entrypoint.sh:101]: I think -socket isn't needed because it defaults to /run/peerpod/hypervisor.sock already.

File: src/cloud-api-adaptor/entrypoint.sh
Comment (beraldoleal) [reply src/cloud-api-adaptor/entrypoint.sh:93]: It does, and we are using in other parts, like: https://github.com/beraldoleal/cloud-api-adaptor/blob/d7578e6beb1e884be63888036a2748ef6347c857/src/cloud-providers/gcp/misc.go#L47.

This is just a basic implementation, for tech-preview, which will have lots of rooms for improvements. I will note this one as an issue once this gets merged, so we can improve latter. 

About the variables, you are right, sending the fix with the next force-push.

File: src/cloud-api-adaptor/entrypoint.sh
Comment (beraldoleal) [reply src/cloud-api-adaptor/entrypoint.sh:101]: I was just playing safe, since others are using it as well. But I will remove with the next force-push.

File: src/cloud-providers/gcp/provider.go
Comment (beraldoleal) [reply src/cloud-providers/gcp/provider.go:72]: Sure, fixed on next force-push.

File: src/cloud-providers/gcp/provider.go
Comment (beraldoleal) [reply src/cloud-providers/gcp/provider.go:95]: Makes sense, noted for a following PR, if you don't mind.

File: src/cloud-providers/gcp/provider.go
Comment (beraldoleal) [reply src/cloud-providers/gcp/provider.go:127]: lol, makes sense. I changed the default to STANDARD. But I will create an issue for future fine tuning.

File: src/cloud-providers/gcp/provider.go
Comment (beraldoleal) [reply src/cloud-providers/gcp/provider.go:138]: I tried to find any reference or documentation and could not find yet. I will keep looking, however:

Network is not created here, only attached, we are using the default one, for now. (To be improved).

With disks I tried simulating several possible situations that would fail, and the boot disk was not there. So I would say it is safe for now.

File: src/cloud-providers/gcp/provider.go
Comment (beraldoleal) [reply src/cloud-providers/gcp/provider.go:189]: Is part of the Provider Interface.

File: src/peerpod-ctrl/Makefile
Comment (beraldoleal) [reply src/peerpod-ctrl/Makefile:66]: accepting only beers. Fixed.

File: src/cloud-providers/gcp/provider.go
Comment (beraldoleal) [reply src/cloud-providers/gcp/provider.go:92]: I would add this to the list, too. If you don't mind.

File: src/cloud-providers/gcp/provider.go
Comment (beraldoleal) [reply src/cloud-providers/gcp/provider.go:126]: @wainersm, Will the PodVM be able to communicate with external clusters within a private network? This is a genuine question.

Given that GKE is not supported at this point, I am concerned it may not add value currently.

**Comment by beraldoleal** [comment]
Hi @wainersm, thanks for your review! Really appreciated!

Let me know if you are ok with the open threads, and even the closed ones, please feel free to reopen.

About the account, I will send you on slack the credentials to a GCP cluster, if you need.

**Comment by beraldoleal** [comment]
@wainersm let me know if you are ok with the changes.

@bpradipt please consider review this when you find some spare cycles.

File: src/cloud-providers/gcp/provider.go
Comment (wainersm) [reply src/cloud-providers/gcp/provider.go:126]: Hi @beraldoleal ! Not expert speaking, but I guess CAA running on an external cluster won't be able to talk to the VM within a private network unless you provide the correct routing.

I'm not saying that you should implement the private IP on this first version. I just wanted to caught your attention to the fact that some providers like AWS provides that flexibility, and you might be interested on doing the same on the follow on PRs.

File: src/cloud-providers/gcp/provider.go
Comment (wainersm) [reply src/cloud-providers/gcp/provider.go:138]: OK. Can be improved and/or fixed later.

**Review by wainersm — APPROVED** [review]
Hi @beraldoleal !

I see that you followed the instructions in https://github.com/confidential-containers/cloud-api-adaptor/blob/main/src/cloud-api-adaptor/docs/addnewprovider.md  which makes me more comfortable to review.

All the comments I made you addressed and/or will on following changes. Unfortunately I don't have GCP account to give it a try but I know that you tested it, so we are fine.

thanks!

**Comment by wainersm** [comment]
@bpradipt @snir911 mind to review too?

**Comment by bpradipt** [comment]
> @bpradipt @snir911 mind to review too?

Yeah, this is in pipeline. Should be doing soon.

File: src/cloud-api-adaptor/install/overlays/gcp/kustomization.yaml
Comment (bpradipt) [inline src/cloud-api-adaptor/install/overlays/gcp/kustomization.yaml:42]: Can env file be used to provide the secrets followed by a local .gitignore file under `overlays/gcp`, similar to aws and azure? This will make it consistent.

https://github.com/confidential-containers/cloud-api-adaptor/blob/main/src/cloud-api-adaptor/install/overlays/aws/kustomization.yaml#L55
https://github.com/confidential-containers/cloud-api-adaptor/blob/main/src/cloud-api-adaptor/install/overlays/azure/kustomization.yaml#L62

File: src/cloud-providers/gcp/provider.go
Comment (bpradipt) [inline src/cloud-providers/gcp/provider.go:1]: Any plans to add unit tests and e2e tests as follow-on PRs ?

File: .gitignore
Comment (bpradipt) [inline .gitignore:41]: Adding the comment on the commit message here.  You can remove the statement on GKE now since 1909 is merged.

Comment (beraldoleal) [reply .gitignore:41]: Are you proposing we bring GKE support back?

If it's okay with you, can we proceed with this version to establish basic GCP support quickly? I can add GKE in a follow-up PR?

File: src/cloud-providers/gcp/provider.go
Comment (beraldoleal) [reply src/cloud-providers/gcp/provider.go:1]: Tests were part of the provisioning code, and we removed from this PR to wait for the GKE fix. But is currently implemented here:

https://github.com/confidential-containers/cloud-api-adaptor/compare/main...beraldoleal:cloud-api-adaptor:gcp-plus-ueda

I just need to check again and enable more test.

So, my plan was to add a follow up PR adding the provisioning + tests after this PR gets merged. Let me know if you agree with that.

File: .gitignore
Comment (bpradipt) [reply .gitignore:41]: I don't see anything which is specific to GKE and the same code will work for GKE (similar to what we have for aws/eks and azure/aks). Hence suggested to drop the statement from the commit message.
Or for GKE there is a need for separate code?

File: src/cloud-providers/gcp/provider.go
Comment (bpradipt) [reply src/cloud-providers/gcp/provider.go:1]: Makes sense to me. The tests can be added as a follow-on PR.

File: src/cloud-api-adaptor/install/overlays/gcp/kustomization.yaml
Comment (beraldoleal) [reply src/cloud-api-adaptor/install/overlays/gcp/kustomization.yaml:42]: Yes.. I considered using env variables for GCP credentials. But GCP typically uses the credentials JSON files, which seems to be the standard among tools, docs and users. I think it might be simpler for users to stick with that. Env files make sense for Azure, but not sure if the same is valid for GCP. Converting each key (specially things like `private_key`) from the GCP credentials to individual variables could be a bit cumbersome.

Let’s discuss if you see it differently.

File: .gitignore
Comment (beraldoleal) [reply .gitignore:41]: Ohh understood now, fixed the commit message. thanks. :)

File: src/cloud-api-adaptor/gcp/README.md
Comment (bpradipt) [inline src/cloud-api-adaptor/gcp/README.md:182]: s/creadentials/credentials

File: src/cloud-providers/gcp/provider.go
Comment (bpradipt) [inline src/cloud-providers/gcp/provider.go:52]: Is having an empty GcpCredentials a valid case?

File: src/cloud-api-adaptor/install/overlays/gcp/kustomization.yaml
Comment (bpradipt) [reply src/cloud-api-adaptor/install/overlays/gcp/kustomization.yaml:42]: Hmm, I see that the gcp code here expects the creds to be a json file. So I think it's ok for now. We can think about this later.

File: src/cloud-providers/gcp/provider.go
Comment (beraldoleal) [reply src/cloud-providers/gcp/provider.go:52]: Yes, depending on the deployment environment, the library can use ADC. I haven't deeply explored it, though.

File: src/cloud-api-adaptor/gcp/README.md
Comment (beraldoleal) [reply src/cloud-api-adaptor/gcp/README.md]: Fixed.

**Review by bpradipt — APPROVED** [review]
/lgtm
Thanks @beraldoleal

**Comment by beraldoleal** [comment]
Thank you @wainersm  and @bpradipt

**Pull request merged**
